### PR TITLE
fix(buffers)!: Count discarded events in received events

### PR DESCRIPTION
**Pull request by jszwedko**
A user was confused about whether `buffer_discarded_events_total`
includes the number of events in `buffer_received_events_total`.

Previously we did not count events as received when they were discarded
due when `on_full` was `drop_newest`. However, I think marking thas as
received and then discarded is better as:

* This better matches `component_discarded_events_total`
  / `component_received_events_total` where events are received by
  components before being discarded
* If we add additional buffer blocking behavior like `drop_oldest` or
  a buffer that drops events older than a TTL then the buffer will have
  received the event before discarding. In this case I think it'd be
  confusing if `drop_newest` was different.

Replaces: https://github.com/vectordotdev/vector/pull/11938

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

File: lib/vector-buffers/src/topology/channel/sender.rs [hunk 1]
```diff
@@ -295,7 +295,11 @@ impl<T: Bufferable> Sink<T> for BufferSender<T> {
             // We've been instructed to drop the next item.
             SendState::DropNext => {
                 if let Some(instrumentation) = this.instrumentation.as_ref() {
-                    if let Some((item_count, _)) = item_sizing {
+                    if let Some((item_count, item_size)) = item_sizing {
+                        instrumentation.increment_received_event_count_and_byte_size(
```
Comment (001wwang) [inline lib/vector-buffers/src/topology/channel/sender.rs:299]: Can we include a note in the upgrade guide for this change?

**Review by 001wwang — COMMENTED** [review]
Sweet, I think the consistency with `component_` metrics is compelling, and good point on future `drop` behavior fitting in with this. I'm for it.

File: lib/vector-buffers/src/topology/channel/sender.rs (hunk 1 above)
Comment (jszwedko) [reply lib/vector-buffers/src/topology/channel/sender.rs:299]: Ah, doh, yes I had that thought and then completely forgot 😅

File: website/content/en/highlights/2022-03-22-0-21-0-upgrade-guide.md [hunk 2]
```diff
@@ -262,6 +263,17 @@ A new `_unmatched` route has now been introduced and the events are no longer di
...
 You can still get the total number of events that match no routes via `component_events_sent_total` with a tag of `output=_unmatched`.
 
+#### `buffer_discarded_events_total` now includes received events {#buffer-discarded-events}
+
+The `buffer_discarded_events_total` now includes all events flowing into
+a buffer, even those discarded when the buffer is full and `drop_newest` is
+configured as the `when_full` behavior.
+
+This brings the metric inline with the component-level received and discarded
+metrics where events are counted as received before discarded as well as leaving
```
Comment (001wwang) [inline website/content/en/highlights/2022-03-22-0-21-0-upgrade-guide.md:273]:
```suggestion
metrics where events are counted as received before being discarded as well as leaving
```

**Review by 001wwang — APPROVED** [review]

**Pull request merged**